RecoMuon/SimMuonAlgos/plugins/MuonSimClassifier.cc

```diff
@@ -0,0 +1,574 @@
+
+    std::vector<int> classif(nmu, 0), ext(nmu, 0);
+    std::vector<int> hitsPdgId(nmu, 0), G4processType(nmu, 0), momPdgId(nmu, 0), gmomPdgId(nmu, 0), momStatus(nmu, 0);
+    std::vector<int> flav(nmu, 0), momFlav(nmu, 0), gmomFlav(nmu, 0), hmomFlav(nmu, 0);
+    std::vector<int> tpId(nmu, -1), tpBx(nmu, 999), tpEv(nmu, 999);
+    std::vector<float> signp(nmu, 0.0), pt(nmu, 0.0), eta(nmu, -99.), phi(nmu, -99.);
+    std::vector<float> prodRho(nmu, 0.0), prodZ(nmu, 0.0), momRho(nmu, 0.0), momZ(nmu, 0.0);
+    std::vector<float> tpAssoQuality(nmu, -1);
+
+    std::unique_ptr<reco::GenParticleCollection> secondaries;     // output collection of secondary muons
+    std::map<TrackingParticleRef, int>         tpToSecondaries; // map from tp to (index+1) in output collection
+    std::vector<int> muToPrimary(nmu, -1), muToSecondary(nmu, -1); // map from input into (index) in output, -1 for null
+    if (linkToGenParticles_) secondaries.reset(new reco::GenParticleCollection());
+
+    // loop on reco muons
+    for(size_t i = 0; i < nmu; ++i) {    
+        edm::RefToBase<reco::Muon> mu = muons->refAt(i);
+	edm::LogVerbatim("MuonSimClassifier") <<"\n reco::Muon # "<<i;
+
+        TrackingParticleRef        tp;
+        edm::RefToBase<reco::Muon> muMatchBack;
+        r2s_it match = recSimColl.find(mu);
+        s2r_it matchback;
+        if (match != recSimColl.end()) {
+            // match->second is vector, front is first element, first is the ref (second would be the quality)
+            tp = match->second.front().first;
+            tpId[i]          = tp.isNonnull() ? tp.key() : -1; // we check, even if null refs should not appear here at all
+            tpAssoQuality[i] = match->second.front().second;
+            s2r_it matchback = simRecColl.find(tp);
+            if (matchback != simRecColl.end()) {
+                muMatchBack = matchback->second.front().first;
+            } else {
+                edm::LogWarning("MuonSimClassifier") << "\n***WARNING:  This I do NOT understand: why no match back? *** \n";
+            }
+        } else if ((trackType_ == reco::GlobalTk) &&
+                    mu->isGlobalMuon()) {
+            // perform a second attempt, matching with the standalone muon
+            r2s_it matchSta = UpdSTA_recSimColl.find(mu);
+            if (matchSta != UpdSTA_recSimColl.end()) {
+                tp    = matchSta->second.front().first;
+                tpId[i]          = tp.isNonnull() ? tp.key() : -1; // we check, even if null refs should not appear here at all
+                tpAssoQuality[i] = matchSta->second.front().second;
+                s2r_it matchback = UpdSTA_simRecColl.find(tp);
+                if (matchback != UpdSTA_simRecColl.end()) {
+                    muMatchBack = matchback->second.front().first;
+                } else {
+                    edm::LogWarning("MuonSimClassifier") << "\n***WARNING:  This I do NOT understand: why no match back in UpdSTA? *** \n";
```

Comment:
how frequent are these LogWarnings?
Also, if they really should not be happening, perhaps make an exception?

Reply:
I don't think such problems are easy to track. An exception is overkill in this case.